File: threadnote-mvp/electron/src/infrastructure/persistence/memory/memoryPipeline.js
import {
  ClaimStatus,
  EntryKind,
  MemoryScope,
  createMemoryRecord
} from "../../../domain/models/threadnoteModels.js";

const SUBSTANTIVE_ENTRY_KINDS = new Set([
  EntryKind.NOTE,
  EntryKind.QUESTION,
  EntryKind.SOURCE
]);

export class MemoryPipeline {
  constructor({ store }) {
    this.store = store;
  }

  buildThreadMemory({ entries = [], claims = [], anchors = [] }) {
    const records = [];

    for (const entry of entries) {
      const working = this.recordWorking(entry);
      if (working) {
        records.push(working);
      }

      const source = this.recordSource(entry);
      if (source) {
        records.push(source);
      }
    }

    for (const claim of claims) {
      const semantic = this.recordSemantic(claim);
      if (semantic) {
        records.push(semantic);
      }
    }

    for (const anchor of anchors) {
      const episodic = this.recordEpisodic(anchor);
      if (episodic) {
        records.push(episodic);
      }
    }

    return records;
  }

  recordWorking(entry) {
    if (!entry?.threadID || !SUBSTANTIVE_ENTRY_KINDS.has(entry.kind)) {
      return null;
    }
    const text = String(entry.summaryText ?? entry.body?.text ?? "").trim();
    if (!text) {
      return null;
    }
    return createMemoryRecord({
      threadID: entry.threadID,
      scope: MemoryScope.WORKING,
      text,
      provenance: `entry:${entry.id}`,
      createdAt: entry.createdAt
    });
  }

  recordEpisodic(anchor) {
    const text = [anchor?.stateSummary, anchor?.coreQuestion].filter(Boolean).join(" -- ");
    if (!anchor?.threadID || !text) {
      return null;
    }
    return createMemoryRecord({
      threadID: anchor.threadID,
      scope: MemoryScope.EPISODIC,
      text,
      provenance: `anchor:${anchor.id}`,
      createdAt: anchor.createdAt
    });
  }

  recordSemantic(claim) {
    if (!claim?.threadID || claim.status !== ClaimStatus.STABLE || !claim.statement) {
      return null;
    }
    return createMemoryRecord({
      threadID: claim.threadID,
      scope: MemoryScope.SEMANTIC,
      text: claim.statement,
      provenance: `claim:${claim.id}`,
      createdAt: claim.updatedAt
    });
  }

  recordSource(entry) {
    if (!entry?.threadID || entry.kind !== EntryKind.SOURCE) {
      return null;
    }
    const title = entry.body?.title ?? entry.summaryText ?? "";
    const url = entry.body?.url ?? entry.sourceMetadata?.locator ?? "";
    const text = [title, url].filter(Boolean).join(" ");
    if (!text) {
      return null;
    }
    return createMemoryRecord({
      threadID: entry.threadID,
      scope: MemoryScope.SOURCE,
      text,
      provenance: `entry:${entry.id}`,
      createdAt: entry.createdAt
    });
  }

  fetchThreadMemory(threadID, scope = null) {
    return this.store.fetchMemoryRecords(threadID, scope);
  }
}

File: threadnote-mvp/Sources/Memory/MemoryPipeline.swift
/// task015–018 — MemoryPipeline.
/// Sync writes for 4 memory scopes. Each write path is a single targeted insert;
/// no LLM or async work in this layer (that lives in M4).

import Foundation

@MainActor
final class MemoryPipeline {

    private let repository: ThreadnoteRepository

    init(repository: ThreadnoteRepository) {
        self.repository = repository
    }

    // MARK: - Working memory (task015)
    // Called on every entry append when entry belongs to a thread.

    func recordWorking(entry: Entry) {
        guard let threadID = entry.threadID else { return }
        // Only record substantive user entries
        let substantiveKinds: Set<EntryKind> = [
            .note, .claim, .question, .evidence, .source,
            .idea, .plan, .decided, .solved, .verified, .dropped
        ]
        guard substantiveKinds.contains(entry.kind) else { return }
        let text = entry.summaryText.isEmpty ? (entry.body.text ?? "") : entry.summaryText
        guard !text.isEmpty else { return }

        let record = MemoryRecord(
            id: UUID(),
            threadID: threadID,
            scope: .working,
            text: text,
            provenance: "entry:\(entry.id.uuidString)",
            createdAt: entry.createdAt
        )
        persist(record)
    }

    // MARK: - Episodic memory (task016)
    // Called when an Anchor is written.

    func recordEpisodic(anchor: Anchor) {
        let parts = [anchor.stateSummary, anchor.coreQuestion]
            .filter { !$0.isEmpty }
        let text = parts.joined(separator: " — ")
        guard !text.isEmpty else { return }

        let record = MemoryRecord(
            id: UUID(),
            threadID: anchor.threadID,
            scope: .episodic,
            text: text,
            provenance: "anchor:\(anchor.id.uuidString)",
            createdAt: anchor.createdAt
        )
        persist(record)
    }

    // MARK: - Semantic memory (task017)
    // Called when a claim reaches a settled status.

    func recordSemantic(claim: Claim) {
        let settledStatuses: Set<ClaimStatus> = [.stable]
        guard settledStatuses.contains(claim.status) else { return }

        let text = claim.statement
        guard !text.isEmpty else { return }

        let record = MemoryRecord(
            id: UUID(),
            threadID: claim.threadID,
            scope: .semantic,
            text: text,
            provenance: "claim:\(claim.id.uuidString)",
            createdAt: claim.updatedAt
        )
        // Replace any existing semantic record for this claim to avoid duplicates
        let existing = repository.fetchMemoryRecords(for: claim.threadID, scope: .semantic)
        if existing.contains(where: { $0.provenance == "claim:\(claim.id.uuidString)" }) {
            return
        }
        persist(record)
    }

    // MARK: - Source memory (task018)
    // Called when a source-kind entry is appended.

    func recordSource(entry: Entry) {
        guard let threadID = entry.threadID else { return }
        guard entry.kind == .source else { return }

        let title = entry.body.title ?? entry.summaryText
        let url = entry.body.url ?? ""
        let text = [title, url].filter { !$0.isEmpty }.joined(separator: " ")
        guard !text.isEmpty else { return }

        let record = MemoryRecord(
            id: UUID(),
            threadID: threadID,
            scope: .source,
            text: text,
            provenance: "entry:\(entry.id.uuidString)",
            createdAt: entry.createdAt
        )
        persist(record)
    }

    // MARK: - Private

    private func persist(_ record: MemoryRecord) {
        repository.insertMemoryRecord(record)
    }
}
